analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -117,5 +118,77 @@ public void IsPattern_OtherPattern_VisitsBothBranches()
+            validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.Should().BeNull());  // ToDo: MMF-2563 should have NotNull instead
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x == null);       // 2x because value has different states
+        }
+
+        [TestMethod]
+        public void RecursivePattern_SetsNotNull_FirstLevel_PreservePreviousConstraint()
+        {
+            const string code = @"
+if (arg is { } value)
+{
+    Tag(""Value"", value);
+    Tag(""ArgNotNull"", arg);
+}
+Tag(""End"", arg);";
+            var setter = new PreProcessTestCheck(OperationKind.ParameterReference, x => x.SetSymbolConstraint(x.Operation.Instance.TrackedSymbol(), TestConstraint.First));
+            var validator = SETestContext.CreateCS(code, ", object arg", setter).Validator;
+            validator.ValidateContainsOperation(OperationKind.RecursivePattern);
+            validator.ValidateTag("Value", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("Value", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("ArgNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeFalse());     // ToDo: MMF-2563 should BeTrue() instead
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x != null && x.HasConstraint(TestConstraint.First));  // 2x because value has different states
+        }
+
+        [TestMethod]
+        public void RecursivePattern_SetsNotNull_Nested()
+        {
+            const string code = @"
+if (arg is Exception { Message: { } msg } ex)
+{
+    Tag(""Msg"", msg);
+    Tag(""Ex"", ex);
+}
+Tag(""End"", arg);";
+            var setter = new PreProcessTestCheck(OperationKind.ParameterReference, x => x.SetSymbolConstraint(x.Operation.Instance.TrackedSymbol(), TestConstraint.First));
+            var validator = SETestContext.CreateCS(code, ", object arg", setter).Validator;
+            validator.ValidateContainsOperation(OperationKind.RecursivePattern);
+            validator.ValidateTag("Msg", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("Msg", x => x.HasConstraint(TestConstraint.First).Should().BeFalse("Constraint from source value should not be propagated to child property"));
+            validator.ValidateTag("Ex", x => x.HasConstraint(TestConstraint.First).Should().BeTrue());
+            validator.ValidateTag("Ex", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.TagValues("End").Should().HaveCount(2).And.OnlyContain(x => x != null && x.HasConstraint(TestConstraint.First));   // 2x because value has different states
+        }
```

Comment:
I would like to see more test cases with discard, const, var and positional patterns, because they are all mentioned in the [recursive pattern section](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/patterns)

I compiled a list of interesting patterns. They should all just work with the current implementation, but are interesting for future reference and to avoid regressions.
[sharplab.io](
https://sharplab.io/#v2:C4LghgzsA0AmIGoA+ABATARgLACgUGYACdQgYUIG9dCbiiUAWQgWQAoBKS62ngNzABOhYIQC8hVgDsApgHdCAeQBGAK2kBjYB2iEAogA8QhGfIPrpAB2ABLAPaSO7ANzceNawDMJI6xEL8hVjAdJXZ2VzcqHDcYwgB6OMIwJMlYQiVCdTBJdOljAFcAG0KI2PjEgDlbEQh8gWkjCwFbCzAlQoBPYUJfY2qC4sIAC2l6gDphMYARDXsoAXzNDkJ7JMJ6j1HpSXNgDos8gCF8kRn1OeAFzUzspMKIW1yUwml9YG2IOxyAW2lgIdssFKtAAvsD3F5WD4/KwApkdHDYGFwVxomVaAl1Ck0mksjklHlJEUSmiYmDSW5PN4en4JBgdAB9TgUPSGSggl76Qgc4AYcIUnhRdEYxKvGl9ERE4oongJbrZNK88X5HaFaSCFHkspUqHiuk6CjMaQQCBgADmDXZhG+EDN3ORAtoQuF5WttvFkn6UpJZS1sR1rHpEjMlhs9nY3s4vQAamBCvlpAAVfIWNUAHmskhgrPMVi+AD4JAyDRzXnz+WVncK5WWPV7iZqIuSQUA=)

Reply:
We already test that parent is not `IsPattern`, that should be enough. The rest are different combinations and it's impossible to test all anyway.